### ISSUE TITLE
MAC base sdk compatibility issue.

### ISSUE THREAD

**Issue by Pankaj003** [google/filament#2876]
Hi,

We are using XCode 9.2 with base sdk MacOS10.13. We are trying to link Filament version 1.5.2 but we are getting compatibility issue with this version. Filament version 1.5.2 needs sdk version 10.15. We can't upgrade our XCode version due to some other library dependency. 
So, could you please let me know what version of Filament release is compatible with Mac OS sdk 10.13.

**Comment by prideout** [comment]
What is the linker error message that you're seeing?

**Comment by romainguy** [comment]
Are you compiling Metal support? I wouldn't be surprised if Metal 2 is what's causing the linkage errors.

**Comment by Pankaj003** [comment]
The error indeed looks like coming from Metal framework. I am setting the backend as OpenGL (filament::Engine::Backend::OPENGL), so shall I remove the Metal framework from the project?

Below are the error details
------------------------------------------------------------------------------------------------------------------------------------------
Undefined symbols for architecture x86_64:
"_objc_opt_new", referenced from:
filament::backend::metal::MetalProgram::MetalProgram(id<MTLDevice>, filament::backend::Program const&) in libbackend.a(MetalHandles.mm.o)
filament::backend::metal::MetalBlitter::compileFragmentFunction(filament::backend::metal::MetalBlitter::BlitFunctionKey) in libbackend.a(MetalBlitter.mm.o)
filament::backend::metal::acquireDrawable(filament::backend::metal::MetalContext*) in libbackend.a(MetalContext.mm.o)
filament::backend::metal::getOrCreateEmptyTexture(filament::backend::metal::MetalContext*) in libbackend.a(MetalContext.mm.o)
filament::backend::metal::PipelineStateCreator::operator()(id<MTLDevice>, filament::backend::metal::PipelineState const&) in libbackend.a(MetalState.mm.o)
filament::backend::metal::DepthStateCreator::operator()(id<MTLDevice>, filament::backend::metal::DepthStencilState const&) in libbackend.a(MetalState.mm.o)
filament::backend::metal::SamplerStateCreator::operator()(id<MTLDevice>, filament::backend::SamplerParams const&) in libbackend.a(MetalState.mm.o)

**Second error**

"_OBJC_CLASS_$_MTLSharedEventListener", referenced from:
objc-class-ref in libbackend.a(MetalDriver.mm.o)

**Comment by romainguy** [comment]
Yes you'd have to compile without Metal support. @bejado do we have a way to do this?

**Comment by Pankaj003** [comment]
Thanks for your reply. I didn't get it, I need to compile my project without linking the Meal framework or build new Filament libs without Metal support?
But I think Filament version 1.5.2 has Metal as one of the dependency library.

**Comment by romainguy** [comment]
You would need to build Filament without its Metal backend. It shouldn't be too hard, and we may have a compile flag for this already. @bejado can advise.

**Comment by Pankaj003** [comment]
Thanks...will wait for your instructions. I am not expert on Mac systems, so need some detail guidance

**Comment by bejado** [comment]
Yes, we have a CMake flag to enable / disable to Metal backend: `FILAMENT_SUPPORTS_METAL`.

The `CMAKE_OSX_DEPLOYMENT_TARGET` flag can be used to force an older SDK version.

I'm able to disable Metal and build Filament for macOS 10.13 with the following:

```
cmake -G Ninja .. -DCMAKE_BUILD_TYPE=Release -DFILAMENT_SUPPORTS_METAL:BOOL=OFF -DCMAKE_OSX_DEPLOYMENT_TARGET=10.13
ninja
```

**Comment by romainguy** [comment]
Thanks Ben! @Pankaj003 Your solution is to build Filament yourself using the instructions from `BUILDING.md` + what Ben just explained.

**Comment by Pankaj003** [comment]
Hi,

Thanks for the build instructions. I tried the cmake command but getting 2 build error(please see attached file filament_1.5.2.rtf
[1.5.2.zip](https://github.com/google/filament/files/5000727/1.5.2.zip)
[1.5.2.zip](https://github.com/google/filament/files/5000731/1.5.2.zip)

)

line: 189
../filament/src/fg/FrameGraphHandle.h:122:13: \cf4 error: \cf2 default member initializer for 'mHandle' needed within definition of enclosing class 'FrameGraphRenderTarget' outside of member functions

line: 210
../filament/src/fg/FrameGraphHandle.h:122:13: \cf4 error: \cf2 default member initializer for 'mLevel' needed within definition of enclosing class 'FrameGraphRenderTarget' outside of member functions

I've downloaded the source code(Source code(tar.gz)) from https://github.com/google/filament/releases/tag/v1.5.2
and run the below commands:

xcode-select --install
export JAVA_HOME="$(/usr/libexec/java_home)"
make -G Ninja .. -DCMAKE_BUILD_TYPE=Release -DFILAMENT_SUPPORTS_METAL:BOOL=OFF -DCMAKE_OSX_DEPLOYMENT_TARGET=10.13 -DCMAKE_INSTALL_PREFIX=../release/filament ../..
ninja

Thanks,
Pankaj

**Comment by bejado** [comment]
We've seen that issue before with old toolchains: https://github.com/google/filament/issues/1294. If you really can't upgrade to a newer Xcode, you can try to manually fix that error. I can't reproduce, but you could start by changing line 122 of FrameGraphHandle.h to:

```
AttachmentInfo() noexcept : mHandle {} { }
```

**Comment by Pankaj003** [comment]
Hi,

@bejado 

We manually fixed the error but now getting other issue(please see attached file Error.png). I checked the issue and it is related to C++14 support. It seems that XCode9.2 don't have full support for C++14. Do you have any suggestion or change apart from upgrading the XCode to higher version?

![Error](https://user-images.githubusercontent.com/8086656/89038859-b842e600-d35e-11ea-9758-81be6ea67439.png)

**Comment by bejado** [comment]
Could you build Filament 1.5.2 with a newer Clang / Xcode, then link against the rest of your project built with Xcode 9.2? I _think_ Clang has ABI compatibility across versions, but I'm not sure.

I did a local build of Filament 1.5.2 (Metal disabled) with Apple clang version 11.0.3 (clang-1103.0.32.62). You could see if that will link successfully against your project.

[filament-1.5.2.zip](https://github.com/google/filament/files/5008498/filament-1.5.2.zip)

**Comment by Pankaj003** [comment]
Hi,
@bejado 
Thanks for the library files. Sorry for the late reply. I used your build library, but getting one linker error

**Undefined symbols for architecture x86_64:
  "filament::Renderer::beginFrame(filament::SwapChain, unsigned long long, void ()(filament::backend::PresentCallable, void*), void*)"**

am I missing any library?

**Comment by bejado** [comment]
Make sure you're compiling against the 1.5.3 headers. The correct method signature is (note the missing `unsigned long long` argument):

```
bool beginFrame(SwapChain* swapChain, backend::FrameFinishedCallback callback = nullptr, void* user = nullptr);
```

The symbol is in the libfilament.a library.

```
$ nm libfilament.a | c++filt | grep 'beginFrame(filament::SwapChain'
0000000000002a50 T filament::Renderer::beginFrame(filament::SwapChain*, void (*)(filament::backend::PresentCallable, void*), void*)
```

**Comment by Pankaj003** [comment]
Thanks @bejado , I fixed the project settings and now it is compiling correctly.

**Comment by Pankaj003** [comment]
Hi @bejado , @romainguy 
I am using the custom build filament-1.5.2.zip(link in above discussion thread). We removed the Metal support from this build as we have some dependency with XCode9.2 and Filament lib was throwing Metal related linker error. With this Filament build my application is showing blank screen on some of the MacOS10.14 and MacOS10.15 systems. I am suspecting it may be some rendering related issue. due to OpenGL. I found that Apple is deprecating OpenGL on Mojave and Catalina OS's.
My question is, will this Filament custom build will work fine on these two OS's? If not then what alternative do I have to use filament with XCode9.2? Is there a way we can enable Metal support with XCode9.2 and Filament?

Thanks,
Pankaj

**Comment by romainguy** [comment]
OpenGL hasn't changed in years on macos, it should work on those two versions. We test daily on macos and we have no issues with the GL backend.

**Comment by bejado** [comment]
Building Filament with Metal requires the latest macOS SDK, but you could potentially modify Filament to build against an older SDK. You can start by searching our codebase for any availability checks (`@available(macOS 10.14, iOS 12, *)`) and comment out any references to newer symbols (such as `MTLSharedEventListener`).

**Comment by Pankaj003** [comment]
Thanks @romainguy and @bejado for your suggestions.
If OpenGL is fine on OSX10.14 and OSX10.15 then there may be some issue in my app itself, I'll check it. 
One more thing, I am using two different swapchains, camera and views for on-screen and off-screen(image save) rendering. While saving the image I am getting black output. Is there a way I can check if the off-screen viewport has correct or some data? I created a test application and there it is working fine but when I integrated the same code with my main Qt based application, off-screen rendering is not working, how can I debug the Filament objects? Do I need to switch the context or something else?

Thanks,
Pankaj

**Comment by bejado** [comment]
Filament maintains its own OpenGL context on a separate thread, so that shouldn't be a problem. How are you creating the off-screen SwapChains? I don't see why Qt would interfere at all.

**Comment by Pankaj003** [comment]
@romainguy , @bejado 

Hi Ben, thanks for your help. I am performing below operations to perform offscreen rendering. 
code snippet from different application calls:
//////////////////////////////////////////////////////////////////////////////////////////////
    filament::Engine* mEngine = nullptr;
    filament::SwapChain* mSwapChain = nullptr;
    filament::SwapChain* mOffscreenSwapChain = nullptr;
    filament::Renderer* mRenderer = nullptr;
    filament::Camera* mMainCamera = nullptr;
    filament::Camera* mOffscreenCamera = nullptr;
    filament::Scene* mScene = nullptr;
    filament::View* mView = nullptr;
    filament::View* mOffscreenView = nullptr;
    utils::Entity mLight;

    auto backend = filament::Engine::Backend::OPENGL;
    mEngine = filament::Engine::create(backend);
    mOffscreenSwapChain = mEngine->createSwapChain(2048, 2048, filament::SwapChain::CONFIG_TRANSPARENT);

    mRenderer = mEngine->createRenderer();
    mOffscreenCamera = mEngine->createCamera();
    mScene = mEngine->createScene();
    mOffscreenView = mEngine->createView();

    mOffscreenView->setClearColor({ 1.0, 0.125, 0.25, 0.0 });
    mOffscreenView->setScene(mScene);

    mOffscreenView->setViewport({0, 0, 2048, 2048});
    mOffscreenView->setCamera(mOffscreenCamera);

void SaveScreenshot(const QString &filepath, bool transparent, int w, int h)
{
    using namespace filament;
    using namespace std::chrono_literals;

    int n_channels = transparent ? 4 : 3;
    auto dataformat = transparent ? backend::PixelBufferDescriptor::PixelDataFormat::RGBA : backend::PixelBufferDescriptor::PixelDataFormat::RGB;
    auto qimageformat = transparent ? QImage::Format_RGBA8888 : QImage::Format_RGB888;
    const Viewport& vp = mOffscreenView->getViewport();

    mCamManipulator.updateCamera(mOffscreenCamera);

    // if transparent, change clear color to full alpha

    auto clear_color = mView->getClearColor();
    if (transparent) {
        clear_color[3] = 0.0;
    } else {
        clear_color[3] = 1.0;
    }
    mOffscreenView->setClearColor(clear_color);

    bool done = false;

    uint8_t* pixels = new uint8_t[vp.width * vp.height * n_channels];
    backend::PixelBufferDescriptor buffer(
        pixels,
        vp.width * vp.height * n_channels,
        dataformat,
        backend::PixelBufferDescriptor::PixelDataType::UBYTE,
        [](void* buffer, size_t size, void *capture_state) {
            CaptureState* state = static_cast<CaptureState*>(capture_state);
            const Viewport &v = state->view->getViewport();
            QImage img(reinterpret_cast<unsigned char*>(buffer), v.width, v.height, state->img_format);
            QImage scaled = img.scaled(state->w, state->h, Qt::IgnoreAspectRatio, Qt::SmoothTransformation);
            scaled.save(state->filepath);

            // cleanup.
            delete[] static_cast<uint8_t*>(buffer);
            delete state;

            *(state->done) = true;
        }, new CaptureState{mOffscreenView, filepath, qimageformat, w, h, &done});

    while (!mRenderer->beginFrame(mOffscreenSwapChain)) {
        std::this_thread::sleep_for(8ms);
    }
    mRenderer->render(mOffscreenView);
    mRenderer->readPixels(0, 0, vp.width, vp.height, std::move(buffer));
    mRenderer->endFrame();

    while (!done) {
        // Keep flushing the render buffer.  
        if (mRenderer->beginFrame(mOffscreenSwapChain)) {
            mRenderer->render(mOffscreenView);
            mRenderer->endFrame();
        }
        std::this_thread::sleep_for(8ms);
    }
}
//////////////////////////////////////////////////////////////////////////////////////////////

output for non transparent image:
![output](https://user-images.githubusercontent.com/8086656/92995184-bacc5a00-f51e-11ea-846e-33d13b84ed4f.png)

**Comment by bejado** [comment]
Where is `mView` set? What does `mView->getClearColor();` return?

**Comment by Pankaj003** [comment]
@bejado , @romainguy 
Hi, 
I am setting mView and other needed variables along with the offscreen view variables. I removed the code while posting it here. getClearColor() call is returning correct rgba values like 1, 0.125, 0.25, 1. I suspect the issue with the compiler clang. I created the sample Qt app and it is using clang from /usr/bin folder and it is working fine.
But XCode is using the clang from toolchain and it may be causing the issue. I tried using the same clang in XCode but it's not picking up(or may be I don't know how to do it).
